sound/soc/sof/sof-audio.c

```diff
@@ -168,6 +168,306 @@ static int sof_widget_setup(struct snd_sof_dev *sdev, struct snd_sof_widget *swi
+	case snd_soc_dapm_aif_out:
+	case snd_soc_dapm_aif_in:
+	case snd_soc_dapm_src:
+	case snd_soc_dapm_asrc:
+	case snd_soc_dapm_siggen:
+	case snd_soc_dapm_effect:
+	case snd_soc_dapm_mux:
+	case snd_soc_dapm_demux:
+		if (widget->dobj.private)
+			return true;
+
+		return false;
+	default:
+		return false;
+	}
+}
+
+static int sof_setup_pipeline_connections(struct snd_sof_dev *sdev,
+					  struct snd_soc_dapm_widget_list *list, int dir)
+{
+	struct snd_soc_dapm_widget *widget;
+	struct snd_soc_dapm_path *p;
+	int ret;
+	int i;
+
+	/*
+	 * Set up connections between widgets in the sink/source paths based on direction.
+	 * Some non-SOF widgets exist in topology either for compatibility or for the
+	 * purpose of connecting a pipeline from a host to a DAI in order to receive the DAPM
+	 * events. But they are not handled by the firmware. So ignore them.
+	 */
+	if (dir == SNDRV_PCM_STREAM_PLAYBACK) {
+		for_each_dapm_widgets(list, i, widget) {
+			if (!is_sof_widget(widget))
+				continue;
+
+			snd_soc_dapm_widget_for_each_sink_path(widget, p)
+				if (is_sof_widget(p->sink)) {
+					ret = sof_route_setup(sdev, widget, p->sink);
+					if (ret < 0)
+						return ret;
+				}
+		}
+	} else {
+		for_each_dapm_widgets(list, i, widget) {
+			if (!is_sof_widget(widget))
+				continue;
+
+			snd_soc_dapm_widget_for_each_source_path(widget, p)
+				if (is_sof_widget(p->source)) {
+					ret = sof_route_setup(sdev, p->source, widget);
+					if (ret < 0)
+						return ret;
+				}
+		}
+	}
+
+	return 0;
+}
+
+static struct snd_sof_widget *sof_find_pipeline_by_id(struct snd_sof_dev *sdev, int pipeline_id)
+{
+	struct snd_sof_widget *swidget;
+
+	list_for_each_entry(swidget, &sdev->widget_list, list)
+		if (swidget->id == snd_soc_dapm_scheduler && swidget->pipeline_id == pipeline_id)
+			return swidget;
+
+	return NULL;
+}
+
+int sof_widget_list_setup(struct snd_sof_dev *sdev, struct snd_sof_pcm *spcm, int dir)
+{
+	struct snd_soc_dapm_widget_list *list = spcm->stream[dir].list;
+	struct snd_soc_dapm_widget *widget;
+	struct snd_sof_widget *w, *_w;
+	int pipe_widget_count = 0;
+	int i, ret;
+	int j;
+
+	/* nothing to set up */
+	if (!list)
+		return 0;
+
+	/* set up widgets in the list */
+	for_each_dapm_widgets(list, i, widget) {
+		struct snd_sof_widget *swidget = widget->dobj.private;
+		struct snd_sof_widget *pipe_widget, *temp_pipe_widget;
+		bool pipeline_found = false;
+
+		if (!swidget)
+			continue;
+
+		ret = sof_widget_setup(sdev, swidget);
+		if (ret < 0)
+			goto widget_free;
+
+		/* find pipeline widget for the pipeline that this widget belongs to */
+		pipe_widget = sof_find_pipeline_by_id(sdev, swidget->pipeline_id);
+		if (!pipe_widget) {
+			ret = -EINVAL;
+			goto widget_free;
+		}
+
+		/* Add it to the spcm pipeline list if it isn't part of it already. */
+		list_for_each_entry(w, &spcm->stream[dir].pipeline_list, list) {
+			if (w->comp_id == pipe_widget->comp_id) {
+				pipeline_found = true;
+				break;
+			}
+		}
```

Comment:
@ranj063  is pipeline widget in this list ? If so can we set up pipeline_list when processing pipeline widget ? Not for every widget

Reply:
Is there only one pipeline for this list ? This function is called for pcm_hw_params.  I assume only one pipeline for this list, so we only need to search pipeline and setup pipeline list once. As you know I want to avoid more list search to reduce delay

Reply:
I checked the code, pipeline widget is not in this list and there maybe two or more pipelines for this list (not very common). As you know there maybe many widgets for complex topology, so searching pipe widget for every widget will cost much time. If pipeline id is smaller than 32 or less, we can use a flag[32] to indicate which pipe is processed.
````
if (flag[pipe id]) continue;
flag[pipe id] = 1;
````